Sprint 15

DmitryBeton opened the pull request:
Реализован слой работы с базой данных с использованием абстракций:
добавлены классы TrackerStore, TrackerCategoryStore, TrackerRecordStore для абстрагирования Core Data;
добавлены Entity для хранения Tracker, TrackerCategory, TrackerRecord;
у каждой Entity есть суффикс, который обозначает, что она используется для хранения в Core Data;
в модели (файле с расширением *.xcdatamodeld) указаны взаимосвязи между Entity;
создан один Persistent Container, который управляет тремя типами Store;
использован NSFetchedResultsController для получения обновлений от Core Data;
можно создавать привычки — и после перезапуска приложения информация отображается; слой хранения данных на базе Core Data выполняет задачи;

DmitryBeton committed:
feat(create-tracker): добавить выбор эмодзи и цвета для трекера

- Добавлены коллекции эмодзи и цветов в CreateTrackerViewController
- Реализована логика выбора эмодзи и цвета с возможностью выбора только одного элемента
- Обновлен интерфейс с использованием UIScrollView для поддержки прокрутки

refactor: перевести расписание на использование WeekDay enum

- Заменен массив Bool на Set<WeekDay> в ScheduleViewController
- Обновлена логика переключения дней недели с использованием enum
- Упрощено создание TrackerSchedule через новый инициализатор
- Обновлен метод фильтрации в TrackerRepositoryMock для работы с WeekDay
- Добавлена конвертация номера дня недели в enum WeekDay

DmitryBeton committed:
refactor: Перевел все с MVP на MVC тк было тяжело внедрять CoreData 🥹

DmitryBeton committed:
feat: создана Core Data модель

- Добавлен файл Tracker.xcdatamodeld
- Созданы три сущности:
  - TrackerCoreData с атрибутами: id, name, colorHex, emoji, scheduleData
  - TrackerCategoryCoreData с атрибутами: id, title
  - TrackerRecordCoreData с атрибутами: id, date, trackerId
- Настроены связи между сущностями:
  - TrackerCategoryCoreData → TrackerCoreData (one to many)
  - TrackerCoreData → TrackerRecordCoreData (one to many)
- Также добавил правила удаления(если что потом сбросить до Nullify):
  - TrackerCategoryCoreData.trackers: Cascade
  - TrackerCoreData.category: Nullify
  - TrackerCoreData.records: Cascade
  - TrackerRecordCoreData.tracker: Nullify

DmitryBeton committed:
feat(create-tracker): Добавлена возможность сохранить&отобразить выбранный цвет и эмоджи

DmitryBeton committed:
refactor(schedule): перевел расписание с структуры TrackerSchedule на enum WeekDay

DmitryBeton committed:
feat: Добавил DataStore, DataProvider и DaysValueTransformer

– DataStore получает данные от CoreData
– DataProvider выступает интерфейсом между CoreData и ViewController
– DaysValueTransformer переводит расписание в удобный для хранения формат

DmitryBeton committed:
feat: добавлен фильтр трекеров по дням

DmitryBeton committed:
feat: Добавлены анимации при отображении трекеров и пустого состояния

DmitryBeton committed:
feat: Добавлена возможность отмечать выполнение трекеров и синхронизация UI
– Реализовано сохранение и удаление выполнения трекера (TrackerRecord) в CoreData
– Исправлена логика переключения состояния кнопки выполнения
– Исправлена синхронизация UI с источником данных (completedRecords)

DmitryBeton committed:
fix(tracker): После создания трекера происходит обновление collectionView

DmitryBeton committed:
feat: logger.info добавлен во все функции

DmitryBeton committed:
refactor: согласно ТЗ сделаны 3 Store & все данные для TrackersVC берутся из dataProvider

DmitryBeton committed:
refactor: вырезал fetchedResultsController.delegate за ненадобностью

DmitryBeton committed:
refactor: убрал закомментированный код перед ревью

DmitryBeton committed:
fix: Если в расписании выбраны все 7 дней, то в ячейке отображается "Каждый день"

NaR0RoW commented on Tracker/CoreData/DataProvider.swift:
```diff
@@ -0,0 +1,171 @@
+    }
+    
+    // Метод для получения предиката фильтрации по текущей дате
+    private func getPredicateForCurrentDate() -> NSPredicate? {
```
Такие комментарии в коде лучше не оставлять. Из названия метода понятно, за что он отвечает. Может быть сейчас немного рано, но советую в будущем прочитать книги Роберта Мартина, а именно "Чистый код" (в целом уже можно начать читать на досуге) и "Чистая архитектура" (лучше после некоторого коммерческого опыта). Конкретно про комментарии можно [тут](https://www.baeldung.com/cs/clean-code-comments) почитать. Это касается большинства комментариев в коде

NaR0RoW commented on Tracker/CoreData/UIColorMarshalling.swift:
```diff
@@ -0,0 +1,42 @@
+        )
+    }
+
+    func color(from hex: String) -> UIColor {
```
Методы можно сделать статическими, чтобы не создавать экземпляр класса

NaR0RoW commented on Tracker/Modules/CreateTracker/Components/ColorCollectionViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+        
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
```
Лучше не использовать `fatalError` в `init(coder:)`, так как это крашит приложение. Есть несколько вариантов решения проблемы:
Вариант 1: Запрещаем инициализацию через `Storyboard/XIB`:
``` swift
@available(*, unavailable)
required init?(coder: NSCoder) {
    nil
}
```

Вариант 2: Если инициализация через `XIB` нужна - реализуем `setup()`:
``` swift
required init?(coder: NSCoder) {
    super.init(coder: coder)
    setup()
}

private func setup() {
    // общая инициализация, если требуется
}
```

NaR0RoW commented on Tracker/Modules/CreateTracker/Components/ColorCollectionViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+    // MARK: - Public Methods
+    func setColor(_ color: UIColor) {
+        view.backgroundColor = color
+        self.layer.borderColor = color.withAlphaComponent(0.3).cgColor
```
`self` везде лучше опускать, если явно того не требуется

NaR0RoW commented on Tracker/Modules/CreateTracker/Components/ColorCollectionViewCell.swift:
```diff
@@ -0,0 +1,59 @@
+            self.layer.borderWidth = 3
+        } else {
+            self.layer.borderWidth = 0
+        }
```
Стоит использовать тернарный оператор:
`layer.borderWidth = isSelected ? 3 : 0`

NaR0RoW commented on Tracker/Modules/CreateTracker/Components/EmojiCollectionViewCell.swift:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    func setSelected(_ isSelected: Bool) {
+        if isSelected {
+            self.backgroundColor = .ypLightGray
+        } else {
+            self.backgroundColor = .clear
+        }
```
По возможности везде следует использовать тернарный оператор

NaR0RoW commented on Tracker/Modules/CreateTracker/CreateTrackerViewController.swift:
```diff
@@ -9,17 +9,46 @@ import UIKit
     private let tableViewItems = ["Категория", "Расписание"]
-    private var selectedSchedule: TrackerSchedule?
+    private let sectionsTitles = ["Emoji", "Цвет"]
+    private let emojiCollectionViewItems = [
```
Все моки лучше хранить в одном месте, например:
``` swift
enum MockData {
    static let emojis = [
        "🙂", "😻", "🌺", "🐶", "❤️", "😱",
        "😇", "😡", "🥶", "🤔", "🙌", "🍔",
        "🥦", "🏓", "🥇", "🎸", "🏝️", "😪"
    ]
}
```

NaR0RoW commented on Tracker/Modules/CreateTracker/CreateTrackerViewController.swift:
```diff
@@ -329,10 +410,145 @@ extension CreateTrackerViewController: UITableViewDelegate {
+                withReuseIdentifier: "emojiCell",
+                for: indexPath
+            ) as? EmojiCollectionViewCell else {
+                fatalError("Unable to dequeue EmojiCollectionViewCell")
```
По своему опыту хотел бы отметить, что в продакшене не используют `fatalError`, советую заменить на `assertionFailure`, [тут подробнее](https://www.hackingwithswift.com/plus/intermediate-swift/understanding-assertions)

NaR0RoW commented on Tracker/Modules/CreateTracker/CreateTrackerViewController.swift:
```diff
@@ -329,10 +410,145 @@ extension CreateTrackerViewController: UITableViewDelegate {
+            ofKind: kind,
+            withReuseIdentifier: TrackerHeaderView.reuseIdentifier,
+            for: indexPath
+        ) as! TrackerHeaderView
```
Не стоит использовать `force` операции

NaR0RoW commented on Tracker/Modules/Trackers/Model/WeekDay.swift:
```diff
@@ -0,0 +1,64 @@
+        case .thursday: return "Четверг"
+        case .friday: return "Пятница"
+        case .saturday: return "Суббота"
+        case .sunday: return "Воскресенье"
```
Давай опускать return в однострочных выражениях

NaR0RoW commented on Tracker/Modules/Trackers/Model/WeekDay.swift:
```diff
@@ -0,0 +1,64 @@
+    case thursday = 4
+    case friday = 5
+    case saturday = 6
+    case sunday = 7
```
Достаточно поставить цифру 1 у первого поля, дальше Swift автоматически проставит в порядке возрастания

NaR0RoW commented on Tracker/Modules/Trackers/TrackersViewController.swift:
```diff
@@ -231,32 +321,51 @@ extension TrackersViewController: UICollectionViewDataSource, UICollectionViewDe
         
-        let tracker = visibleCategories[indexPath.section].trackers[indexPath.item]
-        presenter?.configureCell(cell, with: tracker)
+        configureCell(cell, with: Tracker(id: tracker.id!, name: tracker.name!, color: uiColorMarhalling.color(from: tracker.color!) , emoji: tracker.emoji!))
```
Давай нигде не будем использовать `force` операции

NaR0RoW commented on Tracker/CoreData/DaysValueTransformer.swift:
```diff
@@ -0,0 +1,36 @@
+//
+//  Created by Дмитрий Чалов on 13.12.2025.
+//
+import CoreData
```
Стоит убрать `import CoreData`

DmitryBeton committed:
fix: ⚠️ Критические замечания исправлены:

– 3.5 ❌ Нет ошибочных и грубых force-операций (force unwrap, force cast)
– 5.6 ❌ в проекте нигде, кроме Store.swift и месте создания NSPersistentContainer нет строчки import CoreData; иначе говоря - приложение надежно абстрагировано от Core Data.
– Сделал методы класса UIColorMarshalling статическими
– Использовал тернарный оператор заместо if
– А также другие небольшие изменения
